### PR TITLE
Checking Accept header for supported media types

### PR DESCRIPTION
**Pull request by janbj**

### CODE REVIEW

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 1]
```diff
@@ -108,6 +110,23 @@ public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOExcepti
                                                               expectedHeaders.get(expectedHeader).textValue()))));
     }
 
+    @Test
+    @DisplayName("handleRequest should return UnsupportedMediaType")
+    public void handleRequestShouldReturnUnsupportedMediaType() throws IOException {
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:115]: when should it return UnsupportedMediaType?

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 2]
```diff
@@ -108,6 +110,23 @@ public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOExcepti
                                                               expectedHeaders.get(expectedHeader).textValue()))));
     }
 
+    @Test
+    @DisplayName("handleRequest should return UnsupportedMediaType")
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:114]: remove display names. no reason to keep in sync two things. If you see any, it is once we thought it was a good idea

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 3]
```diff
@@ -108,6 +110,23 @@ public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOExcepti
...
+    public void handleRequestShouldReturnUnsupportedMediaType() throws IOException {
+        Handler handler = new Handler(environment);
+
+        Map<String,String> headers = new ConcurrentHashMap<>();
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:118]: unsupportedHeaders

File: build.gradle [hunk 4]
```diff
@@ -25,7 +25,7 @@ wrapper {
 
 allprojects {
     group = 'com.github.bibsysdev'
-    version = '1.9.0'
+    version = '1.9.1'
```
Comment (axthosarouris) [inline build.gradle:28]: I think the change is a breaking change. I would suggest 1.10.0

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 5]
```diff
@@ -108,6 +110,23 @@ public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOExcepti
...
+        InputStream input = requestWithHeaders(headers);
+
+        ByteArrayOutputStream outputStream = outputStream();
+        handler.handleRequest(input, outputStream, context);
+
+        GatewayResponse<String> response = GatewayResponse.fromOutputStream(outputStream);
+        assertThat(response.getStatusCode(), is(equalTo(UNSUPPORTED_MEDIA_TYPE_STATUS_CODE)));
+        assertThat(response.getBody(), containsString(UNSUPPORTED_ACCEPT_HEADER_VALUE));
+    }
+
```
Comment (axthosarouris) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:129]: I think we should have a test about wildcards and the supported values. A parametrized test could be enough

File: apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java [hunk 6]
```diff
@@ -36,6 +43,31 @@
     protected transient OutputStream outputStream;
     protected transient String allowedOrigin;
 
+    public static final List<String> DEFAULT_SUPPORTED_HEADERS = List.of(APPLICATION_JSON, WILDCARD);
+
+    protected void checkHeaders(RequestInfo requestInfo) throws UnsupportedAcceptHeaderException {
+        if (requestInfo.getHeaders().keySet().contains(HttpHeaders.ACCEPT)) {
+            suppliedAcceptsHeadersAreSupported(requestInfo, listSupportedHeaders());
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java:50]: why is passing a method call as parameter necessary? You can call the private method directly

File: apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java [hunk 7]
```diff
@@ -36,6 +43,31 @@
     protected transient OutputStream outputStream;
     protected transient String allowedOrigin;
 
+    public static final List<String> DEFAULT_SUPPORTED_HEADERS = List.of(APPLICATION_JSON, WILDCARD);
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java:46]: Should we add APPLICATION_JSON_LD?

File: apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java [hunk 8]
```diff
@@ -36,6 +43,31 @@
...
+                .findAny()
+                .orElseThrow(() -> new UnsupportedAcceptHeaderException());
+    }
+
+    protected List<String>  listSupportedHeaders() {
+        return DEFAULT_SUPPORTED_HEADERS;
+    }
+
+    @JacocoGenerated
+    protected String getDefaultResponseHeader() {
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java:67]: defaultHeaders in ApiGatewayHandler should be updated

**Review by axthosarouris — CHANGES_REQUESTED** [review]
The following method should be added to the RequestAuthorizer class.
```
  @Override
    protected void checkHeaders(RequestInfo requestInfo)  {
        //do nothing;
    }
```

File: apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java [hunk 9]
```diff
@@ -36,6 +43,40 @@
...
+                .map(mediaType -> mediaType.withoutParameters())
+                .collect(Collectors.toList());
+    }
+
+    private void suppliedAcceptsHeadersAreSupported(List<MediaType> acceptMediaTypes)
+            throws UnsupportedAcceptHeaderException {
+        acceptMediaTypes.stream()
+                .filter(outer -> listSupportedHeaders().stream()
+                        .filter(inner -> inner.is(outer))
+                        .findAny().isPresent())
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java:67]: Suggestion:
```

    private void suppliedAcceptsHeadersAreSupported(List<MediaType> acceptMediaTypes)
            throws UnsupportedAcceptHeaderException {
        acceptMediaTypes.stream()
                .filter(this::somethingIsHappeningHere)
                .findAny()
                .orElseThrow(UnsupportedAcceptHeaderException::new);
    }

    private boolean somethingIsHappeningHere(MediaType outer) {
        return listSupportedHeaders().stream().anyMatch(inner -> inner.is(outer));
    }
```

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java [hunk 10]
```diff
@@ -36,6 +43,44 @@
...
+        acceptMediaTypes.stream()
+                .filter(acceptMediaType -> inSupportedMediaTypeRange(acceptMediaType))
+                .findAny()
+                .orElseThrow(() -> new UnsupportedAcceptHeaderException());
+    }
+
+    private boolean inSupportedMediaTypeRange(MediaType mediaType) {
+        return listSupportedMediaTypes().stream()
+                .filter(supportedMediaType ->supportedMediaType.is(mediaType))
+                .findAny().isPresent();
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/RestRequestHandler.java:73]: replace with anyMatch()

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java [hunk 11]
```diff
@@ -108,6 +114,47 @@ public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOExcepti
...
+        Handler handler = new Handler(environment);
+
+        Map<String,String> unsupportedHeaders = new ConcurrentHashMap<>();
+        unsupportedHeaders.put(HttpHeaders.ACCEPT, mediaType);
+        InputStream input = requestWithHeaders(unsupportedHeaders);
+
+        ByteArrayOutputStream outputStream = outputStream();
+        handler.handleRequest(input, outputStream, context);
+
+        GatewayResponse<String> response = GatewayResponse.fromOutputStream(outputStream);
```
Comment (brinxmat) [inline apigateway/src/test/java/nva/commons/apigateway/ApiGatewayHandlerTest.java:132]: Extract?

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by axthosarouris — APPROVED** [review]

File: apigateway/src/main/java/nva/commons/apigateway/ContentTypes.java [hunk 12]
```diff
@@ -1,12 +0,0 @@
-package nva.commons.apigateway;
-
-import java.util.Locale;
-
-public final class ContentTypes {
-
-    public static final String APPLICATION_JSON = "application/json".toLowerCase(Locale.ROOT);
-    public static final String APPLICATION_JSON_LD = "application/ld+json".toLowerCase(Locale.ROOT);
```
Comment (axthosarouris) [inline apigateway/src/main/java/nva/commons/apigateway/ContentTypes.java:8]: I think json-ld content type should stay.

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Review by brinxmat — DISMISSED** [review]

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Review by axthosarouris — APPROVED** [review]

**Pull request merged**